tests/device_tests/solana/construct/instructions.py

```diff
@@ -0,0 +1,1443 @@
+        TokenProgramInstruction.INS_INITIALIZE_ACCOUNT_2: TokenProgram_InitializeAccount2_Instruction,
+        TokenProgramInstruction.INS_SYNC_NATIVE: TokenProgram_SyncNative_Instruction,
+        TokenProgramInstruction.INS_INITIALIZE_ACCOUNT_3: TokenProgram_InitializeAccount3_Instruction,
+        TokenProgramInstruction.INS_INITIALIZE_IMMUTABLE_OWNER: TokenProgram_InitializeImmutableOwner_Instruction,
+    },
+)
+
+# Token Program end
+
+# Token 2022 Program begin
+
+
+class Token2022ProgramInstruction(IntEnum):
+    INS_INITIALIZE_ACCOUNT = 1
+    INS_INITIALIZE_MULTISIG = 2
+    INS_TRANSFER = 3
+    INS_APPROVE = 4
+    INS_REVOKE = 5
+    INS_SET_AUTHORITY = 6
+    INS_MINT_TO = 7
+    INS_BURN = 8
+    INS_CLOSE_ACCOUNT = 9
+    INS_FREEZE_ACCOUNT = 10
+    INS_THAW_ACCOUNT = 11
+    INS_TRANSFER_CHECKED = 12
+    INS_APPROVE_CHECKED = 13
+    INS_MINT_TO_CHECKED = 14
+    INS_BURN_CHECKED = 15
+    INS_INITIALIZE_ACCOUNT_2 = 16
+    INS_SYNC_NATIVE = 17
+    INS_INITIALIZE_ACCOUNT_3 = 18
+    INS_INITIALIZE_IMMUTABLE_OWNER = 22
+
+
+Token2022Program_InitializeAccount_Instruction = Struct(
+    "program_index" / Byte,
+    "accounts"
+    / CompactStruct(
+        "account_to_initialize" / Byte,
+        "mint_account" / Byte,
+        "owner" / Byte,
+        "rent_sysvar" / Byte,
+    ),
+    "data"
+    / CompactStruct(
+        "instruction_id" / InstructionIdAdapter(GreedyBytes),
+    ),
+)
+
+Token2022Program_InitializeMultisig_Instruction = Struct(
+    "program_index" / Byte,
+    "accounts"
+    / CompactStruct(
+        "multisig_account" / Byte,
+        "rent_sysvar" / Byte,
+        "signer_accounts" / Byte,
+    ),
+    "data"
+    / CompactStruct(
+        "instruction_id" / InstructionIdAdapter(GreedyBytes),
+        "number_of_signers" / Byte,
+    ),
+)
+
+Token2022Program_Transfer_Instruction = Struct(
+    "program_index" / Byte,
+    "accounts"
+    / CompactStruct(
+        "source_account" / Byte,
+        "destination_account" / Byte,
+        "owner" / Byte,
+        "multisig_signers" / Optional(GreedyRange(Byte)),
+    ),
+    "data"
+    / CompactStruct(
+        "instruction_id" / InstructionIdAdapter(GreedyBytes),
+        "amount" / Int64ul,
+    ),
+)
+
+Token2022Program_Approve_Instruction = Struct(
+    "program_index" / Byte,
+    "accounts"
+    / CompactStruct(
+        "source_account" / Byte,
+        "delegate_account" / Byte,
+        "owner" / Byte,
+        "multisig_signers" / Optional(GreedyRange(Byte)),
+    ),
+    "data"
+    / CompactStruct(
+        "instruction_id" / InstructionIdAdapter(GreedyBytes),
+        "amount" / Int64ul,
+    ),
+)
+
+Token2022Program_Revoke_Instruction = Struct(
+    "program_index" / Byte,
+    "accounts"
+    / CompactStruct(
+        "source_account" / Byte,
+        "owner" / Byte,
+        "multisig_signers" / Optional(GreedyRange(Byte)),
+    ),
+    "data"
+    / CompactStruct(
+        "instruction_id" / InstructionIdAdapter(GreedyBytes),
```

Comment:
you can use `InstructionId` in all those cases, which you helpfully already have defined :)

Reply:
https://github.com/trezor/trezor-firmware/pull/3359/commits/5cc7eab0731883cd9af6fb2e3eba1398a667f12e#diff-3a06ea4817ce485da542a9c3f705d02ecd5f1e7f8da16375d9766218503d4e07R80